airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/buffers/MemoryBoundedLinkedBlockingQueue.java

```diff
@@ -7,77 +7,135 @@
   }
 
   public boolean offer(final E e, final long itemSizeInBytes) {
-    final long newMemoryUsage = currentMemoryUsage.addAndGet(itemSizeInBytes);
-    if (newMemoryUsage <= maxMemoryUsage.get()) {
-      final boolean success = super.offer(new MemoryItem<>(e, itemSizeInBytes));
-      if (!success) {
-        currentMemoryUsage.addAndGet(-itemSizeInBytes);
-      } else {
-        // it succeeded!
-        timeOfLastMessage.set(Instant.now());
-      }
-      log.debug("offer status: {}", success);
-      return success;
-    } else {
-      currentMemoryUsage.addAndGet(-itemSizeInBytes);
-      log.debug("offer failed");
-      return false;
-    }
+    return hiddenQueue.offer(e, itemSizeInBytes);
   }
 
-  @Override
   public MemoryBoundedLinkedBlockingQueue.MemoryItem<E> take() throws InterruptedException {
-    final MemoryItem<E> memoryItem = super.take();
-    if (memoryItem != null) {
-      currentMemoryUsage.addAndGet(-memoryItem.size());
-      return memoryItem;
-    }
-    return null;
+    return hiddenQueue.take();
   }
 
-  @Override
   public MemoryBoundedLinkedBlockingQueue.MemoryItem<E> poll() {
-    final MemoryItem<E> memoryItem = super.poll();
-    if (memoryItem != null) {
+    return hiddenQueue.poll();
+  }
+
+  public MemoryBoundedLinkedBlockingQueue.MemoryItem<E> poll(final long timeout, final TimeUnit unit) throws InterruptedException {
+    return hiddenQueue.poll(timeout, unit);
+  }
+
+  /**
+   * Extends LinkedBlockingQueue so that we can get a LinkedBlockingQueue bounded by memory. Hidden as
+   * an inner class, so it doesn't get misused, see top-level javadoc comment.
+   *
+   * @param <E>
+   */
+  private static class HiddenQueue<E> extends LinkedBlockingQueue<MemoryBoundedLinkedBlockingQueue.MemoryItem<E>> {
+
+    private final AtomicLong currentMemoryUsage;
+    private final AtomicLong maxMemoryUsage;
+    private final AtomicReference<Instant> timeOfLastMessage;
+
+    public HiddenQueue(final long maxMemoryUsage) {
+      currentMemoryUsage = new AtomicLong(0);
+      this.maxMemoryUsage = new AtomicLong(maxMemoryUsage);
+      timeOfLastMessage = new AtomicReference<>(null);
+    }
+
+    public boolean offer(final E e, final long itemSizeInBytes) {
+      final long newMemoryUsage = currentMemoryUsage.addAndGet(itemSizeInBytes);
+      if (newMemoryUsage <= maxMemoryUsage.get()) {
+        final boolean success = super.offer(new MemoryItem<>(e, itemSizeInBytes));
+        if (!success) {
+          currentMemoryUsage.addAndGet(-itemSizeInBytes);
+        } else {
+          // it succeeded!
+          timeOfLastMessage.set(Instant.now());
+        }
+        log.debug("offer status: {}", success);
+        return success;
+      } else {
+        currentMemoryUsage.addAndGet(-itemSizeInBytes);
+        log.debug("offer failed");
+        return false;
+      }
+    }
+
+    @Override
+    public MemoryBoundedLinkedBlockingQueue.MemoryItem<E> take() throws InterruptedException {
```

Comment:
Do we want to make these methods package-private too since `MemoryBoundedLinkedBlockingQueue.MemoryItem<E> poll` is the public-facing side of this queue?

Reply:
looks like we cannot make this package private as the underlying class has this as public and we cannot override the existing signature with a more private visibility.